scipy/special/_logsumexp.py

```diff
@@ -100,53 +101,113 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     """
     xp = array_namespace(a, b)
     a, b = xp_broadcast_promote(a, b, ensure_writeable=True, force_floating=True, xp=xp)
-    axis = tuple(range(a.ndim)) if axis is None else axis
+    a = xp_atleast_nd(a, ndim=1)
+    b = xp_atleast_nd(b, ndim=1) if b is not None else b
 
-    if b is not None:
-        a[b == 0] = -xp.inf
-
-    # Scale by real part for complex inputs, because this affects
-    # the magnitude of the exponential.
     if xp_size(a) == 0:
-        # because `xp.max` doesn't have `initial` argument...
-        shape = np.asarray(a.shape)  # NumPy is concise for scalar or tuple `axis`
+        shape = np.asarray(a.shape)  # NumPy is convenient for shape manipulation
+        axis = tuple(range(a.ndim)) if axis is None else axis
         shape[axis] = 1
-        a_max = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        out = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        sgn = xp.sign(out)
+    else:
+        with np.errstate(divide='ignore', invalid='ignore'):  # log of zero is OK
+            out, sgn = _logsumexp(a, b, axis=axis, return_sign=return_sign, xp=xp)
+
+    # Deal with shape details - reducing dimensions and convert 0-D to scalar for NumPy
+    out = xp.squeeze(out, axis=axis) if not keepdims else out
+    sgn = xp.squeeze(sgn, axis=axis) if (sgn is not None and not keepdims) else sgn
+    out = out[()] if out.ndim == 0 else out
+    sgn = sgn[()] if (sgn is not None and sgn.ndim == 0) else sgn
+
+    return (out, sgn) if return_sign else out
+
+
+def _elements_and_indices_with_max_real(a, axis=-1, xp=None):
+    # We need *an* element with the maximum real part.
```

Comment:
This docstring isn't immediately intuitive to me. What is "*an* element with the maximum real part"? Does that mean, any element that has a real part equal to the maximum found in the array?

Reply:
We need a function that behaves like behaves like np.max with complex input that is Array API Standard compatible.
```python3
np.max([1 + 5j, 2 + 1j, 2 + 0j])  # np.complex128(2+1j)
```
It needs to work for n-d arrays and axis tuples/None.

Reply:
Maybe we need a shim that offers an Array API equivalent of all these functions, so we can just write in NumPy parlance? Such a thing probably already exists. I just imagine it's going to be rather confusing to re-write a bunch of code in ways that handle all the Array API differences, and do that repeatedly for the various occurrences.

Reply:
Yes, this idea has been considered. When I was first introduced to `array_api_compat`, I expected it to do this, and was very surprised that this *wasn't* what it did. See e.g. https://github.com/data-apis/array-api/issues/837#issuecomment-2361816873 for a related idea that has been mulled over for a long time.

But for a while now we have been converting code to use the standard directly, making a few of our own convenience functions where particularly important. 

So whether to have a NumPy to array API shim library or not is a bit beyond this PR; perhaps bring it up in gh-18286 if it hasn't been discussed there explicitly.

Reply:
> ```python
> np.max([1 + 5j, 2 + 1j, 2 + 0j])
> # np.complex128(2+1j), but np.complex128(2+ 0j) would be OK, too. 
> # I don't care about the imaginary part
> ```

would you be okay with spelling such a function like 

```python
max(xp.asarray([1 + 5j, 2 + 1j, 2 + 0j]), complex_ok=True, xp=xp)
```

? If so, since this could be implemented in terms of standard operations, it would be in-scope for https://github.com/data-apis/array-api-extra.

Reply:
> would you be okay with spelling such a function like

Sure, but I would be happy with it just being a `max` that defines a particular ordering for complex numbers; it doesn't need a `complex_ok` arg IMO if the ordering is documented. Thanks for creating the library.